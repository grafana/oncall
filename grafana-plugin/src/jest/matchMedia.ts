// @ts-ignore
export default global.matchMedia =
  global.matchMedia ||
  function (query) {
    return {
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    };
  };
