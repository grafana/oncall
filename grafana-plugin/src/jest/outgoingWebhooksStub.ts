import { OutgoingWebhook } from 'models/outgoing_webhook/outgoing_webhook.types';

export default [
  {
    id: 'K2E45EI2586HS',
    name: 'hook-1',
    team: null,
    webhook: 'http://google.ro',
    data: null,
    user: 'rares',
    password: 'password',
    authorization_header: 'auth-header',
    forward_whole_payload: false,
  },
  {
    id: 'KL3UZQQF2KE5V',
    name: 'hook-3',
    team: null,
    webhook: 'http://google.ro',
    data: null,
    user: null,
    password: null,
    authorization_header: null,
    forward_whole_payload: false,
  },
] as OutgoingWebhook[];
