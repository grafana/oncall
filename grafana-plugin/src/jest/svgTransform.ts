module.exports = {
  process() {
    return { code: 'module.exports = {};' };
  },
  getCacheKey() {
    return 'svgTransform';
  },
};
