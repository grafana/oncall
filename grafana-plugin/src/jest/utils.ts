export function mockUseStore() {
  jest.mock('state/useStore', () => ({
    useStore: () => ({
      isUserActionAllowed: jest.fn().mockReturnValue(true),
    }),
  }));
}
