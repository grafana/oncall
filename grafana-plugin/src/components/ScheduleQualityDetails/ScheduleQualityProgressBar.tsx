import React from 'react';

import { cx } from '@emotion/css';
import { useStyles2 } from '@grafana/ui';
import { bem } from 'styles/utils.styles';

import { getScheduleQualityProgressBarStyles } from './ScheduleQualityProgressBar.styles';

interface ProgressBarProps {
  completed: number;
  className?: string;
  numTotalSteps?: number;
}

export const ScheduleQualityProgressBar: React.FC<ProgressBarProps> = ({ className, completed, numTotalSteps }) => {
  const styles = useStyles2(getScheduleQualityProgressBarStyles);
  const classList = [styles.bar, className || ''];

  return (
    <div className={styles.wrapper}>
      {!numTotalSteps && <div className={classList.join(' ')} style={{ width: `${completed}%` }} />}
      {renderSteps(numTotalSteps, completed)}
    </div>
  );

  function renderSteps(numTotalSteps: number, completed: number) {
    if (!numTotalSteps) {
      return null;
    }

    const maxFillPerRow = 100 / numTotalSteps;
    const rowFill = calculateRowFill(numTotalSteps, completed);

    return new Array(numTotalSteps).fill(0).map((_row, index) => {
      const percentWidth = rowFill[index];

      return (
        <div
          key={index}
          className={cx(styles.row, bem(styles.row, 'progress'))}
          data-testid="progressBar__row"
          style={{ width: `${maxFillPerRow}%` }}
        >
          <div
            className={cx(styles.bar, getClassForCompletionLevel())}
            data-testid="progressBar__bar"
            style={{ width: `${percentWidth}%` }}
          />
        </div>
      );
    });
  }

  function getClassForCompletionLevel() {
    if (completed < 20) {
      return bem(styles.bar, 'danger');
    }
    if (completed < 60) {
      return bem(styles.bar, 'warning');
    }
    return bem(styles.bar, 'primary');
  }

  function calculateRowFill(numTotalSteps: number, completed: number): number[] {
    const fillPerRows = [];
    const maxFillPerRow = 100 / numTotalSteps;
    let leftToFill = completed;

    new Array(numTotalSteps).fill(0).forEach((_value, index) => {
      let currentFill: number;

      currentFill = leftToFill - maxFillPerRow < 0 ? leftToFill : maxFillPerRow;

      leftToFill -= maxFillPerRow;

      let percentWidth = Math.max(0, (currentFill * 100) / maxFillPerRow);
      const shouldSetMinValueInitially = completed > 0 && Math.floor(percentWidth) === 0 && !index;

      if (shouldSetMinValueInitially) {
        percentWidth = 1;
      }

      fillPerRows.push(percentWidth);
    });

    return fillPerRows;
  }
};
