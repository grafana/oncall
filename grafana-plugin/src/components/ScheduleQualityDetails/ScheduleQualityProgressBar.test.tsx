import 'jest/matchMedia';
import React from 'react';

import { render, screen } from '@testing-library/react';

import { ScheduleQualityProgressBar } from './ScheduleQualityProgressBar';

const NUM_STEPS = 5;
const DANGER_CLASS = 'c-progressBar__bar--danger';
const WARNING_CLASS = 'c-progressBar__bar--warning';
const SUCCESS_CLASS = 'c-progressBar__bar--primary';

describe('SourceCode', () => {
  test('It renders 0% complete', () => {
    render(<ScheduleQualityProgressBar completed={0} numTotalSteps={NUM_STEPS} />);

    expect(screen.queryAllByTestId<HTMLElement>('progressBar__bar').length).toEqual(NUM_STEPS);
    const allBars = screen.queryAllByTestId<HTMLElement>('progressBar__bar');
    allBars.forEach((bar) => expect(bar.getAttribute('style').includes('width: 0%')));
  });

  test('It renders 100% complete', () => {
    render(<ScheduleQualityProgressBar completed={100} numTotalSteps={NUM_STEPS} />);

    expect(screen.queryAllByTestId<HTMLElement>('progressBar__bar').length).toEqual(NUM_STEPS);
    const allBars = screen.queryAllByTestId<HTMLElement>('progressBar__bar');
    allBars.forEach((bar) => expect(bar.getAttribute('style').includes('width: 100%')));
  });

  test.each([0, 10, 19])('It renders as danger at <20% completion', (completed) => {
    render(<ScheduleQualityProgressBar completed={completed} numTotalSteps={NUM_STEPS} />);

    screen
      .queryAllByTestId<HTMLElement>('progress__bar')
      .forEach((elem) => expect(Array.from(elem.classList).includes(DANGER_CLASS)));
  });

  test.each([20, 31, 41, 61])('It renders as warning at <60% completion', (completed) => {
    render(<ScheduleQualityProgressBar completed={completed} numTotalSteps={NUM_STEPS} />);

    screen
      .queryAllByTestId<HTMLElement>('progress__bar')
      .forEach((elem) => expect(Array.from(elem.classList).includes(WARNING_CLASS)));
  });

  test.each([60, 61, 79, 99, 100])('It renders as success at >=60% completion', (completed) => {
    render(<ScheduleQualityProgressBar completed={completed} numTotalSteps={NUM_STEPS} />);

    screen
      .queryAllByTestId<HTMLElement>('progress__bar')
      .forEach((elem) => expect(Array.from(elem.classList).includes(SUCCESS_CLASS)));
  });
});
