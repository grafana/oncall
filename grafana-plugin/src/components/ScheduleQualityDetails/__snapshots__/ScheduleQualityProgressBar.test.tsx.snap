// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SourceCode It renders at 0% 1`] = `
<div>
  <div
    class="css-1ixaoku"
  >
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--danger"
        data-testid="progressBar__bar"
        style="width: 0%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--danger"
        data-testid="progressBar__bar"
        style="width: 0%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--danger"
        data-testid="progressBar__bar"
        style="width: 0%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--danger"
        data-testid="progressBar__bar"
        style="width: 0%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--danger"
        data-testid="progressBar__bar"
        style="width: 0%;"
      />
    </div>
  </div>
</div>
`;

exports[`SourceCode It renders at 25% 1`] = `
<div>
  <div
    class="css-1ixaoku"
  >
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--warning"
        data-testid="progressBar__bar"
        style="width: 100%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--warning"
        data-testid="progressBar__bar"
        style="width: 25%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--warning"
        data-testid="progressBar__bar"
        style="width: 0%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--warning"
        data-testid="progressBar__bar"
        style="width: 0%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--warning"
        data-testid="progressBar__bar"
        style="width: 0%;"
      />
    </div>
  </div>
</div>
`;

exports[`SourceCode It renders at 30% 1`] = `
<div>
  <div
    class="css-1ixaoku"
  >
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--warning"
        data-testid="progressBar__bar"
        style="width: 100%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--warning"
        data-testid="progressBar__bar"
        style="width: 50%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--warning"
        data-testid="progressBar__bar"
        style="width: 0%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--warning"
        data-testid="progressBar__bar"
        style="width: 0%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--warning"
        data-testid="progressBar__bar"
        style="width: 0%;"
      />
    </div>
  </div>
</div>
`;

exports[`SourceCode It renders at 50% 1`] = `
<div>
  <div
    class="css-1ixaoku"
  >
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--warning"
        data-testid="progressBar__bar"
        style="width: 100%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--warning"
        data-testid="progressBar__bar"
        style="width: 100%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--warning"
        data-testid="progressBar__bar"
        style="width: 50%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--warning"
        data-testid="progressBar__bar"
        style="width: 0%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--warning"
        data-testid="progressBar__bar"
        style="width: 0%;"
      />
    </div>
  </div>
</div>
`;

exports[`SourceCode It renders at 65% 1`] = `
<div>
  <div
    class="css-1ixaoku"
  >
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--primary"
        data-testid="progressBar__bar"
        style="width: 100%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--primary"
        data-testid="progressBar__bar"
        style="width: 100%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--primary"
        data-testid="progressBar__bar"
        style="width: 100%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--primary"
        data-testid="progressBar__bar"
        style="width: 25%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--primary"
        data-testid="progressBar__bar"
        style="width: 0%;"
      />
    </div>
  </div>
</div>
`;

exports[`SourceCode It renders at 70% 1`] = `
<div>
  <div
    class="css-1ixaoku"
  >
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--primary"
        data-testid="progressBar__bar"
        style="width: 100%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--primary"
        data-testid="progressBar__bar"
        style="width: 100%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--primary"
        data-testid="progressBar__bar"
        style="width: 100%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--primary"
        data-testid="progressBar__bar"
        style="width: 50%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--primary"
        data-testid="progressBar__bar"
        style="width: 0%;"
      />
    </div>
  </div>
</div>
`;

exports[`SourceCode It renders at 100% 1`] = `
<div>
  <div
    class="css-1ixaoku"
  >
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--primary"
        data-testid="progressBar__bar"
        style="width: 100%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--primary"
        data-testid="progressBar__bar"
        style="width: 100%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--primary"
        data-testid="progressBar__bar"
        style="width: 100%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--primary"
        data-testid="progressBar__bar"
        style="width: 100%;"
      />
    </div>
    <div
      class="css-1oyrmju css-1oyrmju--progress"
      data-testid="progressBar__row"
      style="width: 20%;"
    >
      <div
        class="css-qn06wj css-qn06wj--primary"
        data-testid="progressBar__bar"
        style="width: 100%;"
      />
    </div>
  </div>
</div>
`;
